File: src/vision60_map_merger.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String
from geometry_msgs.msg import PoseWithCovarianceStamped
from geometry_msgs.msg import Quaternion
from geometry_msgs.msg import Vector3
from nav_msgs.msg import Odometry
from nav_msgs.msg import OccupancyGrid
from gazebo_msgs.msg import ModelStates
import tf_conversions
from tf import transformations
import tf2_ros
import geometry_msgs.msg
import math

vision60_1_map = None
vision60_2_map = None

def vision60_1_mapCallback(data):
    #vision60_1_map = OccupancyGrid()
    global vision60_1_map
    vision60_1_map = OccupancyGrid()
    # vision60_1_map.header = data.header
    # vision60_1_map.info = data.info
    # x = data.data
    vision60_1_map = data
    #print(vision60_1_map)
    
    #return vision60_1_map.data
    
  
def vision60_2_mapCallback(data):
    #vision60_2_map = OccupancyGrid()
    global vision60_2_map
    vision60_2_map = OccupancyGrid()
    # vision60_2_map.header = data.header
    # vision60_2_map.info = data.info
    # y = data.data
    vision60_2_map = data

    #return vision60_2_map.data

def main():
    global vision60_1_map
    global vision60_2_map

    rospy.init_node('map_merger_node', anonymous=True)
    global_map_pub = rospy.Publisher('/map', OccupancyGrid, queue_size=10)
    rospy.Subscriber("/new_vision60_1/map", OccupancyGrid, vision60_1_mapCallback)
    rospy.Subscriber("/new_vision60_2/map", OccupancyGrid, vision60_2_mapCallback)

    

    while not rospy.is_shutdown():

        vision60_global_map = OccupancyGrid()
        vision60_global_map.header.frame_id = "map"
        vision60_global_map.info.resolution = 0.05
        vision60_global_map.info.origin.position.x =  -10.0
        vision60_global_map.info.origin.position.y = -10.0
        vision60_global_map.info.origin.position.z = 0.0
        vision60_global_map.info.origin.orientation.w = 0.0

        width_ = 384
        height_ = 384
        vision60_global_map.info.width = width_
        vision60_global_map.info.height = height_

        #print(vision60_1_map)
        if vision60_1_map != None and vision60_2_map != None:
            for i in range(len(vision60_1_map.data)):
                #print(i)
                vision60_global_map.data.append(vision60_1_map.data[i] + vision60_2_map.data[i])
                 

                if vision60_global_map.data[i] == -1:
                    vision60_global_map.data[i] = 0
                elif vision60_global_map.data[i] == -2:
                    vision60_global_map.data[i] = -1
                elif vision60_global_map.data[i] == 99 or vision60_global_map.data[i] == 200:
                    vision60_global_map.data[i] = 100

                #print(vision60_global_map.data)
        
            #print("hi")
            # Publish the new maps
            #print(vision60_global_map.data)
            global_map_pub.publish(vision60_global_map)
            

        rate = rospy.Rate(10)  # 10hz
        rate.sleep()
        #rospy.spin()
        


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: src/vision60_map_merger.cpp

#include "ros/ros.h"
#include <nav_msgs/OccupancyGrid.h>
#include <vector>
#include <std_msgs/Header.h>
#include <nav_msgs/MapMetaData.h>

// Define global message 
nav_msgs::OccupancyGrid vision60_1_map, vision60_2_map;


void vision60_1_mapCallback(const nav_msgs::OccupancyGrid::ConstPtr & msg)
{
  vision60_1_map.header = msg->header;
  vision60_1_map.info = msg->info;
  vision60_1_map.data = msg->data;
  std::cout << "Go to vision60_1_map callback" << std::endl;
}

void vision60_2_mapCallback(const nav_msgs::OccupancyGrid::ConstPtr & msg)
{
  vision60_2_map.header = msg->header;
  vision60_2_map.info = msg->info;
  vision60_2_map.data = msg->data;
}

int main(int argc, char * argv[])
{
  ros::init(argc, argv, "map_merger_node");
  ros::NodeHandle nh;

  // Create the publisher and subscriber
  const auto global_map_pub = nh.advertise<nav_msgs::OccupancyGrid>("/map", 100);
  const auto local_map_1_sub = nh.subscribe<nav_msgs::OccupancyGrid>("/new_vision60_1/map", 100, vision60_1_mapCallback);
  const auto local_map_2_sub = nh.subscribe<nav_msgs::OccupancyGrid>("/new_vision60_2/map", 100, vision60_2_mapCallback);

  ros::Rate loop_rate(100);

  while (ros::ok())
  {
    nav_msgs::OccupancyGrid vision60_global_map;
    vision60_global_map.header.frame_id = "map";
    vision60_global_map.info.resolution = 0.05;
    vision60_global_map.info.origin.position.x =  -10.0;
    vision60_global_map.info.origin.position.y = -10.0;
    vision60_global_map.info.origin.position.z = 0.0;
    vision60_global_map.info.origin.orientation.w = 0.0;

    const size_t width_ = 384;
    const size_t height_ = 384;
    vision60_global_map.info.width = width_;
    vision60_global_map.info.height = height_;

    for(int i=0; i<147456; i++)
    {
        vision60_global_map.data[i] = vision60_1_map.data[i] + vision60_2_map.data[i];

        if (vision60_global_map.data[i] == -1)
        {
            vision60_global_map.data[i] = 0;
        }
        else if (vision60_global_map.data[i] == -2)
        {
            vision60_global_map.data[i] = -1;
        }
        else if ((vision60_global_map.data[i] == 99) || (vision60_global_map.data[i] == 200))
        {
            vision60_global_map.data[i] = 100;
        }
    }

    // Publish the new maps
    global_map_pub.publish(vision60_global_map);

    ros::spinOnce();
    loop_rate.sleep();
  }

   return 0; 
}

  
